File: compiler design/lexical_analyser.c
#include <stdio.h>
#include <ctype.h>
#include <string.h>

const char* keywords[] = {"if", "else", "while", "return", "int", "float"};
const int num_keywords = sizeof(keywords) / sizeof(keywords[0]);

int is_keyword(const char* str) {
    for (int i = 0; i < num_keywords; i++) {
        if (strcmp(str, keywords[i]) == 0) return 1;
    }
    return 0;
}

int is_valid_identifier(const char* str) {
    if (!isalpha(str[0]) && str[0] != '_') return 0; 
    for (int i = 1; str[i] != '\0'; i++) {
        if (!isalnum(str[i]) && str[i] != '_') return 0;
    }
    return 1;
}

void get_next_token(const char input[], int *index) {
    char value[100] = {0};
    int i = 0, is_invalid_identifier = 0;
    while (isspace(input[*index])) (*index)++;
    if (input[*index] == '\0') return;

    if (isalpha(input[*index]) || input[*index] == '_') {
        while (isalnum(input[*index]) || input[*index] == '_') value[i++] = input[(*index)++];
        printf("%s: %s\n", is_keyword(value) ? "Keyword" : "Identifier", value);
    } else if (isdigit(input[*index])) {
        while (isalnum(input[*index]) || input[*index] == '_') {
            if (isalpha(input[*index]) || input[*index] == '_') is_invalid_identifier = 1;
            value[i++] = input[(*index)++];
        }
        printf("%s: %s\n", is_invalid_identifier ? "Invalid Identifier" : "Number", value);
    } else if (strchr("+-*/<>!=&|", input[*index])) {
        value[i++] = input[(*index)++];
        if (value[0] == '+' && input[*index] == '+') {
            value[i++] = input[(*index)++];
        } else if (strchr("=<>&|", value[0]) && input[*index] == '=') {
            value[i++] = input[(*index)++];
        }
        printf("Operator: %s\n", value);
    } else if (strchr("(),;{}[]", input[*index])) {
        printf("Delimiter: %c\n", input[(*index)++]);
    } else if (input[*index] == '"') {
        value[i++] = input[(*index)++];
        while (input[*index] != '"' && input[*index] != '\0') value[i++] = input[(*index)++];
        if (input[*index] == '"') value[i++] = input[(*index)++];
        value[i] = '\0';
        printf("String Literal: %s\n", value);
    } else {
        printf("Unknown: %c\n", input[(*index)++]);
    }
}

int main() {
    char input[500];
    int index = 0;

    printf("Enter input string: ");
    fgets(input, sizeof(input), stdin);

    while (input[index] != '\0') {
        get_next_token(input, &index);
    }
    return 0;
}

File: compiler design/lexicalAnalyser.java
import java.util.*;

public class lexicalAnalyser {
    private static final Set<String> keywords = Set.of("for", "if", "else", "while", "return", "int", "float");

    public static boolean isKeyword(String token) {
        return keywords.contains(token);
    }
    
    public static boolean isValidIdentifier(String token) {
        if (!Character.isLetter(token.charAt(0)) && token.charAt(0) != '_') return false;
        for (char ch : token.toCharArray()) {
            if (!Character.isLetterOrDigit(ch) && ch != '_') return false;
        }
        return true;
    }

    public static void analyzeToken(String token) {
        if (token.isEmpty()) return;
        
        if (isKeyword(token)) {
            System.out.println("Keyword: " + token);
        } else if (token.matches("[a-zA-Z\\d+]+=[a-zA-Z\\d+]+")){
            if(token.contains("=")){
                if(token.matches("=")) {System.out.println("Operator: "+token);}
                String[] temp = token.split("(?==)|(?<==)");
                for (String x : temp) {
                    analyzeToken(x);
                }
            }
        } else if (Character.isLetter(token.charAt(0)) || token.charAt(0) == '_') {
            System.out.println(isValidIdentifier(token) ? "Identifier: " + token : "Invalid Identifier: " + token);
        } else if (Character.isDigit(token.charAt(0))) {
            System.out.println(token.matches("\\d+") ? "Number: " + token : "Invalid Identifier: " + token);        
        } else if (token.matches("\\+\\+|--|[+\\-*/<>!=&|]=?")) {
            System.out.println("Operator: " + token);
        } else if (token.matches("[(),;{}\\[\\]]")) {
            System.out.println("Delimiter: " + token);
        } else if (token.startsWith("\"") && token.endsWith("\"") && token.length() > 1) {
            System.out.println("String Literal: " + token);
        } else {
            System.out.println("Unknown: " + token);
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter input string: ");
        String input = scanner.nextLine();
        String[] tokens = input.split("\\s+|"+"(?=[(),;{}\\[\\]])|"+"(?<=[(),;{}\\[\\]])|"
        +"(?=<=|>=|==|!=|\\+\\+|--)|"+"(?<=<=|>=|==|!=|\\+\\+|--)");
        for (String token : tokens) {
            analyzeToken(token);    
        }

        scanner.close();
    }
}